cpp/src/iterator/iterator.cuh

```diff
@@ -0,0 +1,459 @@
+ *
+ * This struct computes the output value as
+ * `static_cast<T>(_value)*static_cast<T>(_value)`.
+ *
+ * This will be used to compute "sum of squares".
+ *
+ * @tparam  T a scalar data type to be output
+ * @tparam  T_input a scalar data type to be input
+ * -------------------------------------------------------------------------**/
+template<typename T>
+struct mutator_squared
+{
+    T value_squared;    /// the value of squared
+
+    template<typename T_input>
+    CUDA_HOST_DEVICE_CALLABLE
+    mutator_squared(T_input _value, bool is_valid=false)
+    {
+        T v = static_cast<T>(_value);
+        value_squared = v*v;
+    };
+
+    CUDA_HOST_DEVICE_CALLABLE
+    mutator_squared(){};
+
+    CUDA_HOST_DEVICE_CALLABLE
+    operator T() const { return value_squared; }
+};
+
+
+/** -------------------------------------------------------------------------*
+ * @brief mutator for output a set of values
+ * A mutator for `column_input_iterator`
+ * `column_input_iterator` creates this struct with parameter `_value`, `is_valid`
+ *
+ * This struct computes the value and the squared value and the count at once
+ * where the count is updated only if update_count = true.
+ *
+ * This is an example case to output a struct from column input.
+ * The result of this will be used to compute mean (= sum / count)
+ * and variance (= sum of squares / count - mean^2).
+ * `update_count=false` will be used when the count is known before computation.
+ *
+ * @tparam  T a scalar data type to be output
+ * @tparam  update_count if true, `count` is updated. else, count` is not updated
+ * @tparam  T_input a scalar data type to be input
+ * -------------------------------------------------------------------------**/
+template<typename T, bool update_count=true>
+struct mutator_meanvar;
+
+// @overload mutator_meanvar<T, true>
+template<typename T>
+struct mutator_meanvar<T, true>
+{
+    T value;                /// the value
+    T value_squared;        /// the value of squared
+    gdf_index_type count;   /// the count
+
+    template<typename T_input>
+    CUDA_HOST_DEVICE_CALLABLE
+    mutator_meanvar(T_input _value, bool is_valid)
+    : value( static_cast<T>(_value) ), count(is_valid? 1 : 0)
+    {
+        value_squared = value*value;
+    };
+
+    CUDA_HOST_DEVICE_CALLABLE
+    mutator_meanvar(T _value, T _value_squared=0, gdf_index_type _count=0)
+    : value(_value), value_squared(_value_squared), count(_count)
+    {};
+
+
+    CUDA_HOST_DEVICE_CALLABLE
+    mutator_meanvar()
+    : value(0), value_squared(0), count(0)
+    {};
+
+    using this_t = cudf::detail::mutator_meanvar<T, true>;
+
+    CUDA_HOST_DEVICE_CALLABLE
+    this_t operator+(this_t const &rhs) const
+    {
+        return this_t(
+            (this->value + rhs.value),
+            (this->value_squared + rhs.value_squared),
+            (this->count + rhs.count)
+        );
+    }
+
+    CUDA_HOST_DEVICE_CALLABLE
+    bool operator==(this_t const &rhs) const
```

Comment:
Why does this `mutator` need an `operator==`, but none of the rest? Make this clear when you document the required interface of every `mutator` class.